crates/wasmtime/src/runtime/ref/gc_ref.rs

```diff
@@ -1,52 +1,318 @@
+/// let mut linker = Linker::new(&engine);
+/// linker.func_wrap(
+///     "host-string",
+///     "new",
+///     |caller: Caller<'_, ()>| -> Rooted<ExternRef> { ExternRef::new(caller, Cow::from("")) },
+/// )?;
+/// linker.func_wrap(
+///     "host-string",
+///     "concat",
+///     |mut caller: Caller<'_, ()>, a: Rooted<ExternRef>, b: Rooted<ExternRef>| -> Result<Rooted<ExternRef>> {
+///         let mut s = a
+///             .data(&caller)?
+///             .downcast_ref::<Cow<str>>()
+///             .ok_or_else(|| Error::msg("externref was not a string"))?
+///             .clone()
+///             .into_owned();
+///         let b = b
+///             .data(&caller)?
+///             .downcast_ref::<Cow<str>>()
+///             .ok_or_else(|| Error::msg("externref was not a string"))?;
+///         s.push_str(&b);
+///         Ok(ExternRef::new(&mut caller, s))
+///     },
+/// )?;
+///
+/// // Here is a Wasm module that uses those APIs.
+/// let module = Module::new(
+///     &engine,
+///     r#"
+///         (module
+///             (import "host-string" "concat" (func $concat (param externref externref)
+///                                                          (result externref)))
+///             (func (export "run") (param externref externref) (result externref)
+///                 local.get 0
+///                 local.get 1
+///                 call $concat
+///             )
+///         )
+///     "#,
+/// )?;
+///
+/// // Create a couple `externref`s wrapping `Cow<str>`s.
+/// let hello = ExternRef::new(&mut store, Cow::from("Hello, "));
+/// let world = ExternRef::new(&mut store, Cow::from("World!"));
+///
+/// // Instantiate the module and pass the `externref`s into it.
+/// let instance = linker.instantiate(&mut store, &module)?;
+/// let result = instance
+///     .get_typed_func::<(Rooted<ExternRef>, Rooted<ExternRef>), Rooted<ExternRef>>(&mut store, "run")?
+///     .call(&mut store, (hello, world))?;
+///
+/// // The module should have concatenated the strings together!
+/// assert_eq!(
+///     result.data(&store)?.downcast_ref::<Cow<str>>().unwrap(),
+///     "Hello, World!"
+/// );
+/// # Ok(())
+/// # }
+/// ```
+#[derive(Debug)]
 #[repr(transparent)]
 pub struct ExternRef {
-    inner: VMExternRef,
+    inner: GcRootIndex,
 }
 
+unsafe impl GcRefImpl for ExternRef {}
+
 impl ExternRef {
     /// Creates a new instance of `ExternRef` wrapping the given value.
-    pub fn new<T>(value: T) -> ExternRef
+    ///
+    /// The resulting value is automatically unrooted when the given `context`'s
+    /// scope is exited. See [`Rooted<T>`][crate::Rooted]'s documentation for
+    /// more details.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use wasmtime::*;
+    /// # fn _foo() -> Result<()> {
+    /// let mut store = Store::<()>::default();
+    ///
+    /// // Create an `externref` wrapping a `str`.
+    /// let externref = ExternRef::new(&mut store, "hello!");
```

Comment:
Technically doesn't this API example leak the `externref` for the entire lifetime of `store`? If so can this perhaps show using a `RootScope` to not do that?